src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -629,111 +630,107 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
+                if (_onCompleted?.Count > 0)
+                {
+                    await FireOnCompleted();
+                }
 
-            // At this point all user code that needs use to the request or response streams has completed.
-            // Using these streams in the OnCompleted callback is not allowed.
-            try
-            {
-                await _bodyControl.StopAsync();
-            }
-            catch (Exception ex)
-            {
-                // BodyControl.StopAsync() can throw if the PipeWriter was completed prior to the application writing
-                // enough bytes to satisfy the specified Content-Length. This risks double-logging the exception,
-                // but this scenario generally indicates an app bug, so I don't want to risk not logging it.
-                ReportApplicationError(ex);
-            }
+                application.DisposeContext(context, _applicationException);
 
-            // 4XX responses are written by TryProduceInvalidRequestResponse during connection tear down.
-            if (_requestRejectedException == null)
-            {
-                if (!_connectionAborted)
+                // Even for non-keep-alive requests, try to consume the entire body to avoid RSTs.
+                if (!_connectionAborted && _requestRejectedException == null && !messageBody.IsEmpty)
                 {
-                    // Call ProduceEnd() before consuming the rest of the request body to prevent
-                    // delaying clients waiting for the chunk terminator:
-                    //
-                    // https://github.com/dotnet/corefx/issues/17330#issuecomment-288248663
-                    //
-                    // This also prevents the 100 Continue response from being sent if the app
-                    // never tried to read the body.
-                    // https://github.com/aspnet/KestrelHttpServer/issues/2102
-                    //
-                    // ProduceEnd() must be called before _application.DisposeContext(), to ensure
-                    // HttpContext.Response.StatusCode is correctly set when
-                    // IHttpContextFactory.Dispose(HttpContext) is called.
-                    await ProduceEnd();
+                    await messageBody.ConsumeAsync();
                 }
-                else if (!HasResponseStarted)
+
+                if (HasStartedConsumingRequestBody)
                 {
-                    // If the request was aborted and no response was sent, there's no
-                    // meaningful status code to log.
-                    StatusCode = 0;
+                    await messageBody.StopAsync();
                 }
-            }
 
-            if (_onCompleted?.Count > 0)
-            {
-                await FireOnCompleted();
+                // Clear any AsyncLocals set during the request; back to a clean state ready for next request
+                ExecutionContext.Restore(cleanContext);
```

Comment:
What happens if an exception occurs prior to this?  Is the thread polluted?

Reply:
Either its caught further up; changing the response to a 4xx, 500 etc
![image](https://user-images.githubusercontent.com/1142958/85185707-03aca380-b28d-11ea-9af9-28046b60d5bb.png)

Or it escapes this method (when the EC.Restore wouldn't run); when the whole connection is burnt and socket is shutdown.

As its running on the ThreadPool the "polluted" thread gets cleaned up when it's returned.